### DISCUSSION TITLE
Trying to use conan emsdk/3.1.49 (which depends on nodejs/16.3.0) with C++ exceptions

### DISCUSSION THREAD

**Discussion by rturrado** [emscripten-core/emscripten#20869] [Q&A]
Hi all,

I have a C++ project that I'm compiling to wasm. I'm using conan as package manager, and I'm requiring the latest `emsdk` conan package, `emsdk/3.1.49`, which is dependant on `nodejs/16.3.0`.

I'm using embind to expose three functions from the C++ project, one of which, when executed, uses C++ exceptions. When running a JS test that calls the exposed API, I'm getting this error:

```
node:internal/process/promises:289
            triggerUncaughtException(err, true /* fromPromise */);
            ^

[UnhandledPromiseRejection: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason "#<Exception>".] {
  code: 'ERR_UNHANDLED_REJECTION'
}

Node.js v20.10.0
```

I'm running the JS test with `node test.js --experimental-wasm-eh`.

I'm using `-fwasm-exceptions` both for compiling and linking the C++ project to wasm.

I wonder if the issue comes from the fact that nodejs support for exception handling isn't available until version 17.0, as stated [here](https://webassembly.org/roadmap/).

If so, do you have any idea when there will be a conan `emsdk` binary dependant on `nodejs/17.0`+?

Many thanks!

**Comment by sbc100** [comment]
From the error message it looks like you are using Node.js v20.10.0, no?

If the C++ code is throwing, is it not expected that you would get and exception like this?   Should you not wrap the call in a try/catch on the C++ side if you expect the code to throw?

Or is the C++ throwing and catching internally?  i.e. are you expecting the exceptions not to escape?

**Reply by rturrado** [reply]
Thanks for your quick reply, @sbc100 .

I'm indeed using Node.js 20.10.0 in my WSL command line. To execute a `test.js`, from where I call some C++ functions in a wasm binary.

But when I require `emsdk/3.1.49` in my `conanfile.py`, I see this `emsdk` depends on a `nodejs/16.3.0` (see [here](https://conan.io/center/recipes/emsdk?version=3.1.49), Dependencies tab).

Regarding the exception, C++ is throwing and catching internally. We are not talking here of uncaught exceptions. So yes, I have like two paths of response, let's say a string representing an AST or a string representing a list of parsing errors. The C++ code throws in case of parsing errors, but catches those throws internally. To the outer world, e.g., to JavaScript code, only a string is sent.

**Reply by sbc100** [reply]
I don't know how conan works, but emsdk internally includes a copy of node 16.3.0 to run the compiler itself.  This is unrelated to whatever version of node you choose to run the generated code under.   That is up to you.

If you are seeing `UnhandledPromiseRejection` then I think you can add an unhandled rejection handler which should give you a stack trace and allow you debug the issue.  See https://www.geeksforgeeks.org/node-js-process-unhandledrejection-event/

**Reply by rturrado** [reply]
OK, I have a few questions, maybe you have the answer to some of them:

- It seems Node.js 16.3.0 doesn't have support exception handling. Does that mean that emsdk 3.1.49, since it has that node 16.3.0 internally, won't have support for exception handling?
- A wasm file without C++ exception support means I cannot use exceptions in that code at all? Or that I cannot throw from C++ to the outside world (e.g., one of the functions I expose to JavaScript cannot throw)?

Anyway, many thanks for the hint and the link. It may be that I'm indeed throwing from my wasm function out, although I'm pretty sure I'm not doing it (but I wouldn't bet my head on it 😉)

**Reply by sbc100** [reply]
You can still build a binary with wasm exceptions enabled using emsdk 3.1.49.  This is because the version of node we use internally is independent of the what is required by the generated code.    If you want to use wasm exceptions you need to run in a more modern version of node, but emcc can still build the binary with 16.3.0 (this is why emsdk internally bundles its version of node and doesn't add it to your PATH if you already have node installed... that way users with a more modern version of node will not be effected by emsdk's choice to use 16.3.0 internally).

If you want to run on an older version of node (it doesn't sounds like that is required in your case but in case it is) you can build with emscripten's old exception handling support (that doesn't depend on wasm exception handling).  See the `-sDISABLE_EXCEPTION_THROWING` and `-sDISABLE_EXCEPTION_CATCHING` settings.

**Reply by rturrado** [reply]
OK, I'm building now with this compile option:
`fwasm-exceptions`.

And these link options:
`-lembind -fwasm-exceptions --profiling-funcs -sASSERTIONS=1 -sEXPORT_EXCEPTION_HANDLING_HELPERS -sMODULARIZE=1`.

I'm processing the `unhandledRejection` error with:
```
process.on('unhandledRejection', error => {
    console.error('unhandledRejection', error, '\n');
});
```

And I'm getting this outpu:
```
unhandledRejection Exception [WebAssembly.Exception] {
  message: [ 'antlr4::InputMismatchException', '' ]
}
```

And that makes absolute sense, because I'm using ANTLR in my code, and I'm parsing a program with syntactic errors, and ANTLR implements its error listeners through exceptions. But all those exceptions are catched in the code. I have C++ tests in my code testing exactly the same parsing of the same program with syntactic errors, and the tests don't throw.

So this is what I still don't understand. This emscripten documentation page about [C++ Exceptions Support](https://emscripten.org/docs/porting/exceptions.html) says that:
1) if you throw-catch in your C++ code, the default wasm behaviour is to terminate the execution, but
2) if you pass `-fwasm-exceptions` both as compile and link flags, that throw-catch C++ code will work just fine.

But I'm doing exactly that, and it's not working for me.

**Reply by sbc100** [reply]
How are you calling into your wasm module?  Is it simply via `main` or some other function?  Can you share the `catch` code that seems to be failing in your case?   We have have extensive tests for try/catch in C++ so it seems odd to me that it would not work in your case.

CC @aheejin who might be able to help.

**Reply by rturrado** [reply]
Sure @sbc100. Many thanks for your help!

I've just simplified the use case a lot.

So this is the C++ function I'm calling, after this last simplification:

```
std::string V3xAnalyzer::parse_string_to_json(const std::string &data, const std::string &filename) {
    (void) data;
    (void) filename;
    try {
        throw std::runtime_error{ "blah" };
    } catch (const std::runtime_error &err) {
        return std::string{ "*** " } + err.what();
    }
    //return cqasm::utils::to_json(v3x::parser::parse_string(data, filename));
}
```

This is the output I'm getting now:

```
unhandledRejection Exception [WebAssembly.Exception] {
  message: [ 'std::runtime_error', 'blah' ]
}
```

This is my JavaScript test program:

```
wrapper = require("../build/Release/emscripten/cqasm_emscripten.js");

process.on('unhandledRejection', error => {
    console.error('unhandledRejection', error, '\n');
});

wrapper().then(function(result) {
    var cqasm = new result["EmscriptenWrapper"]()

    var program_1 = "version 3;qubit[5] q;bit[5] b;h q[0:4];b = measure"
    console.log(
        "\n\nExample 1:", program_1,
        "\n\nCalling parse_string_to_json...",
        "\n\nOutput:", cqasm.parse_string_to_json(program_1, "shor.cq"), "\n"
    )

    cqasm.delete()
});
```

And this is the `emscripten_wrapper.hpp` where I define the binding:

```
#pragma once

#include <emscripten/bind.h>
#include <string>


struct EmscriptenWrapper {
    EmscriptenWrapper();

    std::string parse_string_to_json(const std::string &data, const std::string &filename);
};


EMSCRIPTEN_BINDINGS(CqasmJS) {
    emscripten::class_<EmscriptenWrapper>("EmscriptenWrapper")
    .constructor()
    .function("parse_string_to_json", &EmscriptenWrapper::parse_string_to_json);
}
```

**Reply by rturrado** [reply]
OK. I found the error. My mistake! I'm very sorry!

I was only using the compile/link `-fwasm-exceptions` option for `emscripten_wrapper.cpp`, not for the rest of the code!

I have updated my CMakeLists.txt to use that option for the rest of the code, and everything works as expected now.

Many thanks for your help. Appreciated.

**Comment by sbc100** [comment]
Glad it working for you!

**Reply by rturrado** [reply]
Thanks @sbc100! I still have (at least) one question though.

Let's say I invoke a C++ function from JavaScript. And that funcion executes some third party code, e.g., in my case, the ANTLR runtime binary that I just request through conan and link to my project. If that third party code does throw-catch, since I haven't compiled it myself with `-fwasm-exceptions`, I still have the same problem, haven't I? I.e., that throw-catch code will get to JavaScript as unhandled.

So, would I need to build all my dependencies from sources using `-fwasm-exceptions?

**Reply by sbc100** [reply]
Yes.. every source file that uses exceptions in any way would need to be built with `-fwasm-exceptions`

**Reply by rturrado** [reply]
Perfect. Yes. I did some changes today on my build system to `FetchContent` the ANTLR cpp runtime, and build it using `-fwasm-exceptions` and I'm getting the expected behaviour.

**Reply by rturrado** [reply]
Thanks so much for the help, @sbc100 . From my side, we can close this discussion now, as solved.
